lib/dereference.js

```diff
@@ -28,65 +28,61 @@ function dereference (parser, options) {
  * @param {string} path - The full path of `obj`, possibly with a JSON Pointer in the hash
  * @param {string} pathFromRoot - The path of `obj` from the schema root
  * @param {object[]} parents - An array of the parent objects that have already been dereferenced
- * @param {object[]} processedObjects - An array of all the objects that have already been processed
  * @param {object} dereferencedCache - An map of all the dereferenced objects
  * @param {$Refs} $refs
  * @param {$RefParserOptions} options
  * @returns {{value: object, circular: boolean}}
  */
-function crawl (obj, path, pathFromRoot, parents, processedObjects, dereferencedCache, $refs, options) {
+function crawl (obj, path, pathFromRoot, parents, dereferencedCache, $refs, options) {
   let dereferenced;
   let result = {
     value: obj,
     circular: false
   };
 
-  if (options.dereference.circular === "ignore" || processedObjects.indexOf(obj) === -1) {
```

Comment:
Wait, so you fixed the performance by removing protection against circular references?

Reply:
Nope! 
It's just a redundant check that didn't seem to have any actual impact on perf.

https://github.com/stoplightio/json-schema-ref-parser/blob/55dee7ed978b734a1fd354e01197236b3eaa106c/lib/dereference.js#L108L126 is what actually made impact. That added check didn't contribute.

For a context this is the change I'm talking about https://github.com/stoplightio/json-schema-ref-parser/commit/5c3eece30a212cda80212244d904bd63792e545c
I just didn't notice `options.dereference.circular === "ignore" || processedObjects.indexOf(obj) === -1)` changing anything so I reverted to the initial version of this code (prior to the aforementioned commit)

Reply:
We may potentially just use a bit more conservative change that was implemented here https://github.com/APIDevTools/json-schema-ref-parser/pull/212
It only swapped arrays with sets, but this already mitigated the perf cliff.